### PR TITLE
Add docs for Abstract Infer classes

### PR DESCRIPTION
**Pull request by fehiepsi**
Addresses #1683. My last PR, I don't want to spam PR during the end of the year. Happy new year! :dancer:

### CODE REVIEW

File: pyro/infer/abstract_infer.py [hunk 1]
```diff
@@ -278,4 +302,7 @@ def _traces(self, *args, **kwargs):
             yield (replayed_trace, 0., 0)
 
     def marginal(self, sites=None):
+        """
+        Gets marginal distribution from posterior.
+        """
         return self.posterior.marginal(sites)
```
Comment (fehiepsi) [inline pyro/infer/abstract_infer.py]: not sure why we return marginal of posterior (not predictive) here so I leave it as is.

File: pyro/infer/abstract_infer.py [hunk 2]
```diff
@@ -140,6 +149,12 @@ def support(self, flatten=False):
 
     @property
     def empirical(self):
+        """
+        A dictionary of sites' names and theirs corresponding :class:`EmpiricalMarginal`
```
Comment (jpchen) [inline pyro/infer/abstract_infer.py]: "their"

File: pyro/infer/abstract_infer.py [hunk 3]
```diff
@@ -129,6 +129,15 @@ def _populate_traces(self, trace_posterior, validate):
                            for site in self.sites}
 
     def support(self, flatten=False):
+        """
+        Gets support of this marginal distribution.
+
+        :param bool flatten: A flag to decide if we want to flatten `batch_shape`
+            when the marginal distribution is collected from the posterior with
+            ``num_chains > 1``.
```
Comment (jpchen) [inline pyro/infer/abstract_infer.py]: "defaults to False"

File: pyro/infer/mcmc/mcmc.py [hunk 4]
```diff
@@ -268,11 +268,23 @@ def _traces(self, *args, **kwargs):
...
+        :returns: A :class:`MCMCMarginals` class instance.
+        :rtype: :class:`MCMCMarginals`.
+        """
         return MCMCMarginals(self, sites)
 
 
 class MCMCMarginals(Marginals):
     def diagnostics(self):
+        """
+        Gets some diagnostics statistics such as effective number size and
```
Comment (jpchen) [inline pyro/infer/mcmc/mcmc.py]: "effective sample size" is clearer imo

Comment (fehiepsi) [reply pyro/infer/mcmc/mcmc.py]: Thanks @jpchen! I'll address them soon. :)

**Comment by fritzo** [comment]
@eb8680 can you please confirm that we want to add docs to all these things? I recall you wanted to remove things, and we should avoid adding docs to any classes or methods planned for deprecation.

**Comment by eb8680** [comment]
> confirm that we want to add docs to all these things

Since the methods documented in this PR are already in 0.3, I'm fine with merging this for now.  We can revisit later if we decide to do a cleanup release.

**Pull request merged**